# Walkthrough for iGCSE Computer Science
## Assignment 2 - Programming Task

###  Question 1
As we are asked for an expression we can answer as simply as `a + b` although you may also see these as an assignment expression `z = a + b`

Addition uses the same symbol (which we call an operator as it *operates* on some data) as normal maths:

In [1]:
236+72

308

If we are just asked to divide, we use an 'ordinary' divide, which like in the real world produces a float (decimal) output. However, the divide symbol ÷ is a bit tricky to use - it's not on the keyboard - so we use the / symbol. Think of it like when you do a fraction (which is just division!) Notice here that even though it's a whole number, we still get the .0 to show it's a float.

In [2]:
3500/70

50.0

Cubed is a bit tricky - we can't easily <sup>superscript</sup> text when we type into an ordinary IDE or interpreter, so instead we need some operator to mean "to the power of". Python uses ** - * is multiplication and if you think about it, using powers is like **super** multiplication! So it gets two *'s!

In [3]:
5**3

125

The remainder is known as the modulus (or modulo). We use the % operator - I remember it's this one because the circles on the % look like o's and modulo ends in an o!

In [4]:
39%7

4

### Question 2
Although the question here says 'expression' - we really need to write a little program. 

`*Write an expression to turn a Celsius temperature into Fahrenheit. Begin by declaring a variable
called celtemp and assign it an integer value of your choice.
Then declare a variable called fahtemp, set up so that it takes the value of celtemp, divides it by
5, multiplies by 9, and adds 32.
For a final line, write a print statement for fahtemp showing the result.*`

Well let's do this one line at a time shall we?

*Begin by declaring a variable called celtemp and assign it an integer value of your choice*

- Well we declare a variable by writing it's name, `celtemp`. 
- We then need to use the assignment operator `=` to give it a value.
- Finally we need to give it an actual value - in this case an integer which is a whole number. Let's say, 36?

In [5]:
celtemp = 36

*Then declare a variable called fahtemp, set up so that it takes the value of celtemp, divides it by
5, multiplies by 9, and adds 32.*

So again we have to:
- Declare a variable by writing it's name, `fahtemp`.
- Use the assignment operator `=` to state that we're going to set `fahtemp` to something.
- Write out the sum "celtemp divided by 5 multiplied by 9 added to 32" - using our operators this is `celtemp / 5 * 9 + 32`
- As a note of point, you may see this written as a formula including a fraction of 9/5 - just like in ordinary maths we can include brackets to force the order of operation - in this case it would be `(celtemp * (9/5))+32` but with the way our sum is written it's not needed.

In [6]:
fahtemp = celtemp / 5 * 9 + 32

*For a final line, write a print statement for fahtemp showing the result.*

print statements are really easy. We used them in assignment one, let's just think about what they are for a minute: print is a *function* - that is a piece of code which does some work for us, that we can call by name. It's name is `print`. The way we can *tell* it's a function is those brackets afterwards - `print()`.

But what are the brackets for? Well, to be able to do the work for us print needs be be given one or more pieces of information which we call parameters. I like to think of the brackets as cuddle arms - the function is opening it's arms and it brings all the bits it needs to do it's work in for a big cuddle!

Now the parameters. The easiest way we can output things is to put all the things we want to output inside the cuddle arms. These can be variables such as `fahtemp` and `celtemp`, or they can be bits of text inside `'quotations'` - and we separate them with commas. Python will stick them all together, replacing the commas with spaces.

So we want to print out something like "36 C is the same as 96.8 F" - the numbers are variables so our parameters will be `celtemp`, 'C is the same as', `fahtemp`, 'F'. 

In [7]:
print(celtemp,'C is the same as',fahtemp,'F')

36 C is the same as 96.8 F


### Question 3

Once again we will take this question one line step at a time.

*Accept the price of an item as input. This will include both pounds and pence (or dollars and
cents, or your local currency). At this stage, we will not worry about errors, for example
accepting more than two digits in the pence/cents part of the amount;*

Now on the face of it this looks really easy.
- Declare a variable name.
- Use the assignment operator because we're going to be changing that variable.
- Use the `input` function to get some input from the user.
- Include a parameter inside the cuddle arms to give the user a message.

(It's not this easy! But we'll come back to that in a minute!)

In [8]:
price = input('Please enter the prices of the item: ')

Please enter the prices of the item: 7.48


*Accept as input the total amount of cash given. For example, for an amount costing £7.48, a
customer may hand over a £10 note;*

And again this should be exactly the same as the previous input, right?

In [9]:
paid = input('Please enter the money paid: ')

Please enter the money paid: 10


*Calculate and print out the amount of change due.*

- So we define a variable, `change`
- Use the assignmenbt operator, `=`
- Write out the expression, `paid - price`
- Use the print statement to output some text and the change

In [10]:
change = paid - price
print('Your change is',change)

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Uh-oh! There's a problem there isn't ther? 

Look at the last line of the error - it say's `TypeError` and then tells us that the operand `-` doesn't work for strings!

When we type something into the keyboard it is stored as text (a string) **even if** it looks like a number. To convert it to a number we need to **typecast** it. In this case, because there are decimal points, we have to typecast it to a `float`.

Luckily typecasting is pretty easy. we use a function with the same name as the *new* type (float) and for the parameters (inside the cuddle arms) we put the *old* variable. 

`floatNumber = float(stringNumber)`

Even better, we can include the output of one function as the input (parameter) of another, so":

`floatNumber = float(input('enter a number'))`

So we need to write that lot out again, including the floats typecasts.

In [11]:
price = float(input('Please enter the prices of the item: '))
paid = float(input('Please enter the money paid: '))
change = paid - price
print('Your change is',change)

Please enter the prices of the item: 1.50
Please enter the money paid: 10.00
Your change is 8.5


Well that's not bad, is it? However, we'd get back 8.50 rather than 8.5, so we want to format that change. Python has a `format` function. There's quite a bit of information on the [input/output manual](https://docs.python.org/3/tutorial/inputoutput.html) page but I'll give you the basics:
- We use the `format` function.
- The *first* parameter is the variable we want to format.
- The *second* parameter is a string which describes the format we want to use.

In this case we want to format the part of the number *after* the decimal point to 2dp - so we use `.2` and we want to format it as a float so we have an `f` - so we use `.2f` as the formatting string. Memorise this one as it is probably the most common formatting string you'll come across!

In [12]:
print("Your change is",format(change,".2f"))

Your change is 8.50
